process/block/preprocess/transactionsV2.go

```diff
@@ -0,0 +1,416 @@
+	}
+
+	miniBlocks := txs.getMiniBlockSliceFromMapV2(mapMiniBlocks, mapSCTxs)
+	scheduledMiniBlocks := txs.createScheduledMiniBlocks(haveTime, isShardStuck, isMaxBlockSizeReached, sortedTxs, mapSCTxs)
+	miniBlocks = append(miniBlocks, scheduledMiniBlocks...)
+
+	log.Debug("createAndProcessMiniBlocksFromMeV2",
+		"self shard", txs.shardCoordinator.SelfId(),
+		"gas consumed in sender shard", gasConsumedByMiniBlocksInSenderShard,
+		"total gas consumed in self shard", totalGasConsumedInSelfShard)
+
+	for _, miniBlock := range miniBlocks {
+		log.Debug("mini block info",
+			"type", miniBlock.Type,
+			"sender shard", miniBlock.SenderShardID,
+			"receiver shard", miniBlock.ReceiverShardID,
+			"gas consumed in receiver shard", mapGasConsumedByMiniBlockInReceiverShard[miniBlock.ReceiverShardID],
+			"txs added", len(miniBlock.TxHashes),
+			"not sc txs", mapTxsForShard[miniBlock.ReceiverShardID],
+			"sc txs", mapScsForShard[miniBlock.ReceiverShardID])
+	}
+
+	log.Debug("createAndProcessMiniBlocksFromMeV2 has been finished",
+		"total txs", len(sortedTxs),
+		"num txs added", numTxsAdded,
+		"num txs bad", numTxsBad,
+		"num txs failed", numTxsFailed,
+		"num txs skipped", numTxsSkipped,
+		"num txs with initial balance consumed", numTxsWithInitialBalanceConsumed,
+		"num cross shard sc calls or special txs", numCrossShardScCallsOrSpecialTxs,
+		"used time for computeGasConsumed", totalTimeUsedForComputeGasConsumed,
+		"used time for processAndRemoveBadTransaction", totalTimeUsedForProcesss)
+
+	return miniBlocks, nil
+}
+
+func (txs *transactions) getMiniBlockSliceFromMapV2(
+	mapMiniBlocks map[uint32]*block.MiniBlock,
+	mapSCTxs map[string]struct{},
+) block.MiniBlockSlice {
+	miniBlocks := make(block.MiniBlockSlice, 0)
+
+	for shardID := uint32(0); shardID < txs.shardCoordinator.NumberOfShards(); shardID++ {
+		if miniBlock, ok := mapMiniBlocks[shardID]; ok {
+			if len(miniBlock.TxHashes) > 0 {
+				miniBlocks = append(miniBlocks, splitMiniBlockIfNeeded(miniBlock, mapSCTxs)...)
+			}
+		}
+	}
+
+	if miniBlock, ok := mapMiniBlocks[core.MetachainShardId]; ok {
+		if len(miniBlock.TxHashes) > 0 {
+			miniBlocks = append(miniBlocks, splitMiniBlockIfNeeded(miniBlock, mapSCTxs)...)
+		}
+	}
+
+	return miniBlocks
+}
+
+func splitMiniBlockIfNeeded(miniBlock *block.MiniBlock, mapSCTxs map[string]struct{}) block.MiniBlockSlice {
+	splitMiniBlocks := make(block.MiniBlockSlice, 0)
+	notScTxHashes := make([][]byte, 0)
+	scTxHashes := make([][]byte, 0)
+
+	for _, txHash := range miniBlock.TxHashes {
+		_, isSCTx := mapSCTxs[string(txHash)]
+		if !isSCTx {
+			notScTxHashes = append(notScTxHashes, txHash)
+			continue
+		}
+
+		scTxHashes = append(scTxHashes, txHash)
+	}
+
+	if len(notScTxHashes) > 0 {
+		notScMiniBlock := &block.MiniBlock{
+			TxHashes:        notScTxHashes,
+			SenderShardID:   miniBlock.SenderShardID,
+			ReceiverShardID: miniBlock.ReceiverShardID,
+			Type:            miniBlock.Type,
+			Reserved:        miniBlock.Reserved,
```

Comment:
do we really need the field Reserved ? Couldn't we integrate in miniBlock.Type ?

Reply:
I would not change the mini block data structure in this PR. This should be discussed.